org.scalafmt.sbt.ScalafmtSbtReporter$ScalafmtSbtError: scalafmt: [v3.7.16] corrupted class path

For example: https://github.com/typelevel/toolkit/actions/runs/6830956374/job/18579714429?pr=122#step:6:25

But this seems to be widespread 🤔

The `org.scalafmt.interfaces.ScalafmtException: [v3.7.16] corrupted class path:` exception is caused by following when running `scalafmt` cli tool with scalafmt config set to `3.7.16`

```
Caused by: java.lang.ClassNotFoundException: org.scalafmt.config.Config
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:592)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at org.scalafmt.dynamic.ScalafmtReflect.<init>(ScalafmtReflect.scala:21)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$.$anonfun$loadClassPath$1(ScalafmtModuleLoader.scala:73)
	at scala.util.Try$.apply(Try.scala:210)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$.org$scalafmt$dynamic$ScalafmtModuleLoader$$loadClassPath(ScalafmtModuleLoader.scala:71)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$WithDownloader.$anonfun$load$2(ScalafmtModuleLoader.scala:37)
	at scala.util.Success.fold(Try.scala:281)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$WithDownloader.load(ScalafmtModuleLoader.scala:37)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$CachedProxy.load$1(ScalafmtModuleLoader.scala:56)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$CachedProxy.$anonfun$load$3(ScalafmtModuleLoader.scala:57)
	at scala.util.Try$.apply(Try.scala:210)
	at org.scalafmt.dynamic.utils.ReentrantCache.getOrAddToCache(ReentrantCache.scala:41)
	at org.scalafmt.dynamic.ScalafmtModuleLoader$CachedProxy.load(ScalafmtModuleLoader.scala:57)
	at org.scalafmt.dynamic.ScalafmtConfigLoader$.$anonfun$load$1(ScalafmtConfigLoader.scala:34)
	at scala.util.Either.flatMap(Either.scala:352)
	at org.scalafmt.dynamic.ScalafmtConfigLoader$.load(ScalafmtConfigLoader.scala:32)
	at org.scalafmt.dynamic.ScalafmtConfigLoader$CachedProxy.load$1(ScalafmtConfigLoader.scala:74)
	at org.scalafmt.dynamic.ScalafmtConfigLoader$CachedProxy.$anonfun$load$12(ScalafmtConfigLoader.scala:76)
	at scala.util.Try$.apply(Try.scala:210)
	at org.scalafmt.dynamic.utils.ReentrantCache.getOrAddToCache(ReentrantCache.scala:41)
	at org.scalafmt.dynamic.ScalafmtConfigLoader$CachedProxy.$anonfun$load$7(ScalafmtConfigLoader.scala:76)
	at scala.util.Success.fold(Try.scala:281)
	at org.scalafmt.dynamic.ScalafmtConfigLoader$CachedProxy.load(ScalafmtConfigLoader.scala:70)
	at org.scalafmt.dynamic.ScalafmtDynamic.resolveConfig(ScalafmtDynamic.scala:59)
	at org.scalafmt.dynamic.ScalafmtDynamic.createSession(ScalafmtDynamic.scala:53)
	at org.scalafmt.cli.ScalafmtDynamicRunner$.run(ScalafmtDynamicRunner.scala:28)
```

--- a/build.sbt
+++ b/build.sbt
@@ -80,6 +80,7 @@ lazy val dynamic = project
     scalacOptions ++= scalacJvmOptions.value
   )
   .dependsOn(interfaces)
+  .dependsOn(core.jvm % "test")
   .enablePlugins(BuildInfoPlugin)
 
 lazy val interfaces = project

--- a/scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtDynamic.scala
+++ b/scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtDynamic.scala
@@ -16,9 +16,11 @@ final case class ScalafmtDynamic(
   def this() = this(
     ScalafmtProperties(),
     new ScalafmtModuleLoader.CachedProxy(
-      new ScalafmtModuleLoader.WithDownloader(CoursierDependencyDownloader)
+      ScalafmtDynamic.defaultUncachedModuleLoader
     ),
-    new ScalafmtConfigLoader.CachedProxy(ScalafmtConfigLoader)
+    new ScalafmtConfigLoader.CachedProxy(
+      ScalafmtDynamic.defaultUncachedConfigLoader
+    )
   )
 
   override def clear(): Unit = {
@@ -57,5 +59,14 @@ final case class ScalafmtDynamic(
 
   def resolveConfig(configPath: Path): FormatEval[ScalafmtReflectConfig] =
     configLoader.load(configPath, properties, moduleLoader)
+
+}
+
+private[dynamic] object ScalafmtDynamic {
+
+  def defaultUncachedModuleLoader =
+    new ScalafmtModuleLoader.WithDownloader(CoursierDependencyDownloader)
+
+  def defaultUncachedConfigLoader = ScalafmtConfigLoader
 
 }

--- a/scalafmt-dynamic/src/test/scala/org/scalafmt/dynamic/DynamicSuite.scala
+++ b/scalafmt-dynamic/src/test/scala/org/scalafmt/dynamic/DynamicSuite.scala
@@ -5,7 +5,7 @@ import java.nio.charset.StandardCharsets
 import java.nio.file.{Files, Path, Paths}
 import java.nio.file.attribute.FileTime
 
-import org.scalafmt.interfaces.{PositionException, Scalafmt, ScalafmtReporter}
+import org.scalafmt.interfaces.{PositionException, ScalafmtReporter}
 import PositionSyntax._
 
 import scala.collection.mutable
@@ -66,12 +66,14 @@ class DynamicSuite extends FunSuite {
           )
         }
       }
-    val dynamic: ScalafmtDynamic = cfgFunc(
-      Scalafmt
-        .create(this.getClass.getClassLoader)
-        .withReporter(reporter)
-        .asInstanceOf[ScalafmtDynamic]
-    )
+    val dynamic: ScalafmtDynamic = cfgFunc {
+      val configLoader = ScalafmtDynamic.defaultUncachedConfigLoader
+      new ScalafmtDynamic(
+        properties = ScalafmtProperties(reporter = reporter),
+        moduleLoader = new ScalafmtModuleLoader.CachedProxy(getModuleLoader),
+        configLoader = new ScalafmtConfigLoader.CachedProxy(configLoader)
+      )
+    }
     val config = Files.createTempFile("scalafmt", ".scalafmt.conf")
     val filename = Paths.get(name + ".scala")
     var timestamps = 100L
@@ -83,7 +85,7 @@ class DynamicSuite extends FunSuite {
     def setVersion(newVersion: String, dialect: String, rest: String*): Unit = {
       val dialectLine = Option(dialect).fold("")(x => s"runner.dialect = $x")
       setConfig(s"""
-        |version=$newVersion
+        |version="$newVersion"
         |$dialectLine
         |${rest.mkString("\n")}
         |""".stripMargin)
@@ -171,6 +173,7 @@ class DynamicSuite extends FunSuite {
   }
 
   private val testedVersions = Seq(
+    nightly,
     latest,
     "3.1.2",
     "2.7.5",
@@ -202,8 +205,6 @@ class DynamicSuite extends FunSuite {
     }
   }
 
-  def latest = BuildInfo.previousStable
-
   def checkVersion(version: String, dialect: String): Unit = {
     check(s"v$version") { f =>
       f.setVersion(version, dialect)
@@ -211,6 +212,7 @@ class DynamicSuite extends FunSuite {
     }
   }
 
+  checkVersion(nightly, "scala212")
   checkVersion(latest, "scala212")
   checkVersion("1.5.1", "scala211")
   checkVersion("1.0.0", "scala211")
@@ -585,7 +587,24 @@ class DynamicSuite extends FunSuite {
 
 private object DynamicSuite {
 
+  def nightly = BuildInfo.nightly
+  def latest = BuildInfo.previousStable
+
   def getDialectError(version: String, dialect: String) =
     if (version >= "3.1.0") s" [dialect $dialect]" else ""
+
+  // in tests, let's not try to download current version
+  def getModuleLoader = new ScalafmtModuleLoader {
+    private val downloader = ScalafmtDynamic.defaultUncachedModuleLoader
+    private val clsLoader = org.scalafmt.Scalafmt.getClass.getClassLoader
+    override def load(
+        cfg: Path,
+        ver: ScalafmtVersion,
+        props: ScalafmtProperties
+    ): FormatEval[ScalafmtReflect] =
+      if (ver.toString == nightly) Right(ScalafmtReflect(clsLoader, ver))
+      else downloader.load(cfg, ver, props)
+    override def close(): Unit = downloader.close()
+  }
 
 }

